src/traces/scattergl/index.js

```diff
@@ -556,21 +556,24 @@ function sceneUpdate(gd, subplot) {
         scene.draw = function draw() {
             var i;
             for(i = 0; i < scene.count; i++) {
-                if(scene.fill2d) scene.fill2d.draw(i);
+                if(scene.fill2d && scene.fillOptions[i]) {
+                    // must do all fills first
+                    scene.fill2d.draw(i);
```

Comment:
Clear improvement over `master` - but this does raise the question of whether this (not your change, but the structure surrounding it) is really the right behavior, particularly with `fill: 'tonext'` but also with multiple fills to zero. I can try this out tomorrow and see if my interpretation of what this is doing is accurate, but In svg `scatter`, each fill gets drawn *below* any trace that it's attached to (even the trace before it if it's a `fill: 'tonext'`) but *above* any other trace that comes before it. If all the lines and points are above all the fills, you can get points and lines that look disconnected from their own fills.